### DISCUSSION TITLE
Why it's safe to switch to the new configuration at apply time？

### DISCUSSION THREAD

**Discussion by themanforfree** [etcd-io/etcd#16796] [Q&A] [answered]
Suppose the initial state has three nodes, S1 S2 S3, S1 is current leader.
at t1, and they all use the C1 config.
at t2, user propose a new config C2 to add S4. S1 append C2 to S2 and S3, than leader commit and use C2, S4 start and use C2.
at t3, user propose a new config C3 to add S5. S1 append C3 to S2, S3 and S4, than leader commit and use C3, S4 commit and use C3, S5 start and use C3, S2 and S3 have log entry of C2 C3, but not commited, they always use C1.
In this case, nodes that use C1 and nodes that use C2 have two disjoint majorities, two leaders may be elected. Will this happen in etcd, and if not, what is the reason?
| node | t1  | t2  | t3  |
| ---- | --- | --- | --- |
| S1   | C1  | C2  | C3  |
| S2   | C1  | C1  | C1  |
| S3   | C1  | C1  | C1  |
| S4   |     | C2  | C3  |
| S5   |     |     | C3  |

**Comment by serathius** [comment, marked as answer]
Raft supports multiple config changes at once, etcd does not.

**Reply by themanforfree** [reply]
The above behavior is to make single-node configuration changes twice in a row. After S1 commits C2, S1 can start to process C3, but S2 S3 may not committed C2 yet.
